kentik_api_library/kentik_api/analytics/flatness.py

```diff
@@ -0,0 +1,280 @@
+import io
+import json
+import logging
+import sys
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import pandas as pd  # type: ignore
+
+from kentik_api.utils import DeviceCache  # type: ignore
+
+from .data_frame_cache import DFCache  # type: ignore
+
+log = logging.getLogger("flatness_analysis")
+
+
+@dataclass
+class Interval:
+    start: datetime
+    end: datetime
+
+
+class FlatnessResults:
+    """
+    Class storing results of traffic flatness analysis on set of links
+    The events attribute contains dictionary keyed by link name containing list intervals in which traffic was
+    deemed to be 'flat'. Each interval is instance of 'Interval' named tuple.
+    """
+
+    def __init__(self, links: List[str]):
+        self.events: Dict[str, List[Interval]] = {link: [] for link in links}
+
+    def __getitem__(self, link):
+        return self.events.get(link)
+
+    @property
+    def stats(self) -> Dict[str, int]:
+        return dict(
+            total_events=sum(len(i) for i in self.events.values()),
+            max_per_link=max(len(i) for i in self.events.values()),
+        )
+
+    def last(self, link: str) -> Optional[Interval]:
+        if link in self.events and self.events[link]:
+            return self.events[link][-1]
+        else:
+            return None
+
+    def set_last(self, link: str, interval: Interval) -> None:
+        self.events[link][-1] = interval
+
+    @property
+    def affected_links(self) -> List[str]:
+        return [link for link, intervals in self.events.items() if len(intervals) > 0]
+
+    def to_json(self, filename: Optional[str] = None) -> Optional[str]:
```

Comment:
I think it would be nice to split this method into to_json_file and to_json_string

Reply:
I've not found any case in which I'd need the JSON string without eventually writing it to file. The `to_text` method provides reasonably human consumable output. In most cases, the application using the framework will use the data to produce UI specific rendering of results. I'll soon add Jupyter notebook showing an example.

Reply:
The optional filename works as a secret switch to alter the function behavior. 
But, It won't kill anybody to have it like this, I guess

Reply:
You're right. Secret/magic is to be avoided. I'll split it to `to_json` and `to_json_file`.